### ISSUE TITLE
Add symfony 7 support to vendor-plugin and recipe-plugin

### ISSUE THREAD

**Issue by emteknetnz** [silverstripe/.github#325]
silverstripe/vendor-plugin:^2 and silverstripe/recipe-plugin:^2 both have some method signatures that assume symfony 6 is being used

CMS 6 uses symfony 7 and there don't appear to be any issues with this in the CI builds for CMS 6. However we've had community feedback that there are issues is some circumstances with this in [this issue](https://github.com/silverstripe/recipe-plugin/issues/46) and [this issue](https://github.com/silverstripe/vendor-plugin/issues/82).

It probably makes sense to release v3 versions for both of these modules to be released that have method signatures updated to match symfony 7, and to update any CMS 6 modules that use these modules accordingly

> [!IMPORTANT]
> See https://github.com/silverstripe/vendor-plugin/issues/82#issuecomment-2646778662 for more context including how this actually ends up happening.
> tl;dr it's possible to install composer with different dependency versions, which means we may need to allow `^2|^3` constraints for CMS 5 as well??

## Related issues
- https://github.com/silverstripe/recipe-plugin/issues/46
- https://github.com/silverstripe/vendor-plugin/issues/82

## Notes

- Offline discussion concludes that there is minimal risk by making these changes in a minor release, since the likelyhood of anyone subclassing and overriding these methods is extremely low and this doesn't break anything for people using Symfony 6 because more specific return types are permitted.

## Acceptance criteria
- [ ] ~~New major version of silverstripe/recipe-plugin and silverstripe/vendor-plugin are released~~
- [ ] This code change is made as an exception to our [definition of public API](https://docs.silverstripe.org/en/project_governance/public_api/) i.e. it is done in a minor release
- [ ] Return types are updated to match those in symfony 7

## Pull Requests

- https://github.com/silverstripe/recipe-plugin/pull/49
- https://github.com/silverstripe/vendor-plugin/pull/89
- https://github.com/silverstripe/developer-docs/pull/685

**Comment by GuySartorelli** [comment]
> CMS 5 modules remain on ^2

Given this is about what version of the module is used in _composer_, it might be sensible for us to have a `^2|^3` constraint for CMS 5. That way it will be forward compatible with later composer versions.

**Comment by emteknetnz** [comment]
I'd like to avoid dual support as it caused a lot of maintenance problems in CMS 4.

The "used in composer" part of this I'm finding fairly confusing. I'm making the assumption here that composer means the composer/composer dependency that's installed with silverstripe i.e. in a CMS 6 kitchen-sink installed e.g. `composer show | grep composer/composer` will show `2.8.1` on my local.

I don't think `^2|^3` for CMS 5 makes any sense? What good would the ^3 do for say silverstripe/vendor-plugin ([required](https://github.com/silverstripe/silverstripe-framework/blob/5/composer.json#L42) by silverstripe/framework)? CMS 5 requires symfony 6.  If silverstripe/vendor-plugin:^3 is type hinted for symfony 7, then won't that then cause an issue since it's being installed with symfony 6?

**Comment by GuySartorelli** [comment]
> I'm making the assumption here that composer means the composer/composer dependency that's installed with silverstripe

No, I mean literally composer, as in when you run `composer require silverstripe/framework`. Composer is a PHP project that has PHP dependencies.

**Comment by GuySartorelli** [comment]
> I don't think ^2|^3 for CMS 5 makes any sense? What good would the ^3 do for say silverstripe/vendor-plugin ([required](https://github.com/silverstripe/silverstripe-framework/blob/5/composer.json#L42) by silverstripe/framework)? CMS 5 requires symfony 6. If silverstripe/vendor-plugin:^3 is type hinted for symfony 7, then won't that then cause an issue since it's being installed with symfony 6?

The dependencies required _by_ silverstripe's various modules are entirely irrelevant. The plugin needs to be compatible with the dependencies that the installed version of composer _on the user's host machine_ is using when they run `composer vendor-expose` or `composer require`, etc.

**Comment by emteknetnz** [comment]
> The dependencies required by silverstripe's various modules are entirely irrelevant

Are you sure? Are you saying that https://github.com/silverstripe/silverstripe-framework/blob/5/composer.json#L42 is irrelevant? If so, why it is even in frameworks composer.json?

**Comment by emteknetnz** [comment]
> The plugin needs to be compatible with the dependencies that the installed version of composer on the user's host machine is using when they run composer vendor-expose or composer require, etc.

Well, it's certainly a confusing topic

My understanding is that composer is installed as a downloaded phar which is over 100K lines long and has all its own composer dependencies i.e. the entire vendor folder "baked in". So that's what the dependencies on the host machine are?

My local is running PHP 8.1 (for CMS 5 compatibility) and composer 2.8.1, when I installed composer did I get a version with symfony 6 support? Or symfony 7 support? Do I simply download the pre-baked [composer.phar](https://github.com/composer/composer/releases/download/2.8.1/composer.phar) hosted on the [2.8.1 release page](https://github.com/composer/composer/releases/tag/2.8.1)? Or does it dynamically generate one based on my local environment? Symfony 7 requires PHP 8.2.  I can run composer vendor-expose with no issues

In CI we're using shivammathur/setup-php which will happily install PHP 8.3 with symfony 7 and run a bunch `composer` commands with no issue

If what you're saying is correct about "The plugin needs to be compatible with the dependencies that the installed version of composer on the user's host machine" - it seems that putting `^2|^3` for vendor-plugin in silverstripe/framework composer.json will only take into account what silverstripe/framework wants, not what the host machine wants?

**Comment by madmatt** [comment]
The linked two issues were closed, but in case it's relevant I also ran into this issue on Platform.sh. Weirdly, both my local version of `composer` (installed as a phar) and the one on Platform.sh (installed as individual modules inside `/app/.global/composer` and symlinked to `/usr/local/bin/composer` seemed to include `symfony/console` version 7 as far as I could tell, but only the one on Platform.sh seems to break.

I'm happy to show the issue and let someone poke around an example codebase if it would help with resolving the issue or confirming that a fix works as expected if that's helpful, just hit me up on Slack 🙂

**Comment by GuySartorelli** [comment]
I've been able to test this using the following steps. I did this in a docker container so I could just throw it away afterward.

1. Make sure you're using PHP >=8.2.
2. Make new directory for composer. We're going to clone into it.
3. Swap to that new directory
4. Run the following, which will install composer (using composer ironically) from scratch, and replace `composer` in your CLI with the new installation:
    ```sh
    git clone git@github.com:composer/composer.git .
    git checkout 2.8.5
    composer config --unset platform.php
    rm composer.lock
    composer install
    # note the below may be different for you - use `which composer` before hand to check the actual path.
    rm /usr/local/bin/composer
    ln -s "${PWD}/bin/composer" /usr/local/bin/composer
    ```
5. Then either install a fresh SS installation, or if you already have one, swap to it.
6. Run `composer vendor-expose`.

**Comment by madmatt** [comment]
Ah nice, I'll also try this on Platform.sh today and get back to you

**Comment by emteknetnz** [comment]
@madmatt Have you had a chance to test this on platform.sh?

**Comment by GuySartorelli** [comment]
@emteknetnz You should be able to test yourself with the instructions I provided above.

**Comment by emteknetnz** [comment]
I'd say this should work.  I tested a slightly different way

I created a docker container with a base 6.0 installer with two PRs included

In my docker container in the `~` dir, I installed composer using the standard instructions https://getcomposer.org/download/, though I updated this line to use an old version of composer - `php composer-setup.php --version="2.2.0"`

I validated the signature of the composer execute() method did NOT have the `: int` strong return type via `nano composer.phar` and "where" (search) `n execute(`

I then ran `php ~/composer.phar vendor-expose` - things seemed to work correctly

**Comment by madmatt** [comment]
Just to confirm - this works for me too on Platform.sh. Removing the `dependencies` block customising Platform.sh's version of composer.json and then updating the version of `vendor-plugin` installed in the project from `^2` to `2.x-dev` correctly runs `composer vendor-expose` without complaint.

Thanks team! When should I expect a stable release of the `vendor-plugin` module - the next minor release?

**Comment by GuySartorelli** [comment]
Yup, it'll be in CMS 5.4.0 in April
